{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BERT-fy CORD-19 data\n",
    "\n",
    "Original code from https://www.kaggle.com/theamrzaki/covid-19-bert-researchpapers-semantic-search#Data-Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import json\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "dataset_size = 'small'\n",
    "all_data_path = './../data/'\n",
    "data_path = f'./../data/{dataset_size}/'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "517912"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_json = glob.glob(f'{all_data_path}/**/*.json', recursive=True)\n",
    "\n",
    "len(all_json)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# get only rows with attached files\n",
    "\n",
    "metadata_path = f'{all_data_path}/metadata.csv'\n",
    "stripped_metadata_path = f'{all_data_path}/stripped_metadata.csv'\n",
    "\n",
    "if not os.path.exists(stripped_metadata_path):\n",
    "    meta_df = pd.read_csv(metadata_path, dtype={\n",
    "        'pubmed_id': str,\n",
    "        'Microsoft Academic Paper ID': str,\n",
    "        'doi': str\n",
    "    })\n",
    "\n",
    "    stripped_meta_df = meta_df.dropna(subset=['pmc_json_files'])\n",
    "\n",
    "    stripped_meta_df.to_csv(stripped_metadata_path)\n",
    "\n",
    "    stripped_meta_df.head()\n",
    "\n",
    "    del stripped_meta_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "\n",
    "# create shuffled subset of metadata rows\n",
    "\n",
    "small_metadata_path = f'{all_data_path}/small/metadata.csv'\n",
    "if not os.path.exists(small_metadata_path):\n",
    "    small_metadata_file = open(f'{all_data_path}/small/metadata.csv', 'w')\n",
    "    # get header from metadata.csv\n",
    "    print(subprocess.run(\n",
    "        ['head', '-n 1', f'{all_data_path}/stripped_metadata.csv'], stdout=small_metadata_file))\n",
    "    #!head -n 1 \"{all_data_path}/stripped_metadata.csv\" > \"{all_data_path}/metadata.csv\"\n",
    "    # get random sample from metadata.csv\n",
    "    print(subprocess.run(\n",
    "        ['shuf', '-n 12500', f'{all_data_path}/stripped_metadata.csv'], stdout=small_metadata_file))\n",
    "    #!shuf -n 12500 \"{all_data_path}/stripped_metadata.csv\" >> \"{root_path}/metadata.csv\"\n",
    "    small_metadata_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>cord_uid</th>\n",
       "      <th>sha</th>\n",
       "      <th>source_x</th>\n",
       "      <th>title</th>\n",
       "      <th>doi</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>license</th>\n",
       "      <th>abstract</th>\n",
       "      <th>publish_time</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>mag_id</th>\n",
       "      <th>who_covidence_id</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>pdf_json_files</th>\n",
       "      <th>pmc_json_files</th>\n",
       "      <th>url</th>\n",
       "      <th>s2_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>627783</td>\n",
       "      <td>sep45sga</td>\n",
       "      <td>04d0cf9898016ae1295e1bca17a6b3a5fde89802</td>\n",
       "      <td>Elsevier; Medline; PMC</td>\n",
       "      <td>The absence of coronavirus in expressed prosta...</td>\n",
       "      <td>10.1016/j.reprotox.2020.06.006</td>\n",
       "      <td>PMC7286226</td>\n",
       "      <td>32534021</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Due to the cellular entry of the novel coronav...</td>\n",
       "      <td>2020-06-10</td>\n",
       "      <td>Zhang, Shiqi; Wang, Xiaobo; Zhang, Hong; Xu, A...</td>\n",
       "      <td>Reprod Toxicol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>document_parses/pdf_json/04d0cf9898016ae1295e1...</td>\n",
       "      <td>document_parses/pmc_json/PMC7286226.xml.json</td>\n",
       "      <td>https://api.elsevier.com/content/article/pii/S...</td>\n",
       "      <td>219560235.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>807648</td>\n",
       "      <td>doj39520</td>\n",
       "      <td>bc35bf4d3debb1af7ff8dc285d06f822fab24a7e</td>\n",
       "      <td>Medline; PMC</td>\n",
       "      <td>SARS-CoV-2 receptor ACE2 expression in the hum...</td>\n",
       "      <td>10.1093/eurheartj/ehaa410</td>\n",
       "      <td>PMC7239191</td>\n",
       "      <td>32383758</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-05-14</td>\n",
       "      <td>Thum, Thomas</td>\n",
       "      <td>Eur Heart J</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>document_parses/pdf_json/bc35bf4d3debb1af7ff8d...</td>\n",
       "      <td>document_parses/pmc_json/PMC7239191.xml.json</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pubmed/32383758/;...</td>\n",
       "      <td>218557404.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>611676</td>\n",
       "      <td>lsfkpd5l</td>\n",
       "      <td>8ff5c024beea4259b7b1b0e143bbd6a1734ccfef; 9039...</td>\n",
       "      <td>Medline; PMC</td>\n",
       "      <td>PARP9-DTX3L ubiquitin ligase targets host hist...</td>\n",
       "      <td>10.1038/ni.3279</td>\n",
       "      <td>PMC4653074</td>\n",
       "      <td>26479788</td>\n",
       "      <td>no-cc</td>\n",
       "      <td>Enhancing the response to interferon could off...</td>\n",
       "      <td>2015-12-01</td>\n",
       "      <td>Zhang, Yong; Mao, Dailing; Roswit, William T; ...</td>\n",
       "      <td>Nat Immunol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>document_parses/pdf_json/8ff5c024beea4259b7b1b...</td>\n",
       "      <td>document_parses/pmc_json/PMC4653074.xml.json</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/pubmed/26479788/;...</td>\n",
       "      <td>19031101.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>643594</td>\n",
       "      <td>hjriid4h</td>\n",
       "      <td>8a61e2af9bdd4ba0b83becf314017f3f3a8b9adf</td>\n",
       "      <td>Medline; PMC</td>\n",
       "      <td>Towards defining core principles of public hea...</td>\n",
       "      <td>10.1186/s12889-020-09307-y</td>\n",
       "      <td>PMC7527265</td>\n",
       "      <td>32998729</td>\n",
       "      <td>cc-by</td>\n",
       "      <td>BACKGROUND: European Member States, the Europe...</td>\n",
       "      <td>2020-10-01</td>\n",
       "      <td>Belfroid, Evelien; RoÎ²kamp, Dorothee; Fraser, ...</td>\n",
       "      <td>BMC Public Health</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>document_parses/pdf_json/8a61e2af9bdd4ba0b83be...</td>\n",
       "      <td>document_parses/pmc_json/PMC7527265.xml.json</td>\n",
       "      <td>https://doi.org/10.1186/s12889-020-09307-y; ht...</td>\n",
       "      <td>222071121.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>745390</td>\n",
       "      <td>pczdapvd</td>\n",
       "      <td>7448479341487a112b8e4f87c117622a861eeb0b; ef74...</td>\n",
       "      <td>Medline; PMC</td>\n",
       "      <td>The UPTAKE study: a cross-sectional survey exa...</td>\n",
       "      <td>10.1136/bmjopen-2021-048856</td>\n",
       "      <td>PMC8210694</td>\n",
       "      <td>34130964</td>\n",
       "      <td>cc-by-nc</td>\n",
       "      <td>OBJECTIVE: A key challenge towards a successfu...</td>\n",
       "      <td>2021-06-15</td>\n",
       "      <td>Sethi, Sonika; Kumar, Aditi; Mandal, Anandadee...</td>\n",
       "      <td>BMJ Open</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>document_parses/pdf_json/7448479341487a112b8e4...</td>\n",
       "      <td>document_parses/pmc_json/PMC8210694.xml.json</td>\n",
       "      <td>https://doi.org/10.1136/bmjopen-2021-048856; h...</td>\n",
       "      <td>235450913.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  cord_uid                                                sha  \\\n",
       "0      627783  sep45sga           04d0cf9898016ae1295e1bca17a6b3a5fde89802   \n",
       "1      807648  doj39520           bc35bf4d3debb1af7ff8dc285d06f822fab24a7e   \n",
       "2      611676  lsfkpd5l  8ff5c024beea4259b7b1b0e143bbd6a1734ccfef; 9039...   \n",
       "3      643594  hjriid4h           8a61e2af9bdd4ba0b83becf314017f3f3a8b9adf   \n",
       "4      745390  pczdapvd  7448479341487a112b8e4f87c117622a861eeb0b; ef74...   \n",
       "\n",
       "                 source_x                                              title  \\\n",
       "0  Elsevier; Medline; PMC  The absence of coronavirus in expressed prosta...   \n",
       "1            Medline; PMC  SARS-CoV-2 receptor ACE2 expression in the hum...   \n",
       "2            Medline; PMC  PARP9-DTX3L ubiquitin ligase targets host hist...   \n",
       "3            Medline; PMC  Towards defining core principles of public hea...   \n",
       "4            Medline; PMC  The UPTAKE study: a cross-sectional survey exa...   \n",
       "\n",
       "                              doi       pmcid pubmed_id   license  \\\n",
       "0  10.1016/j.reprotox.2020.06.006  PMC7286226  32534021     no-cc   \n",
       "1       10.1093/eurheartj/ehaa410  PMC7239191  32383758     no-cc   \n",
       "2                 10.1038/ni.3279  PMC4653074  26479788     no-cc   \n",
       "3      10.1186/s12889-020-09307-y  PMC7527265  32998729     cc-by   \n",
       "4     10.1136/bmjopen-2021-048856  PMC8210694  34130964  cc-by-nc   \n",
       "\n",
       "                                            abstract publish_time  \\\n",
       "0  Due to the cellular entry of the novel coronav...   2020-06-10   \n",
       "1                                                NaN   2020-05-14   \n",
       "2  Enhancing the response to interferon could off...   2015-12-01   \n",
       "3  BACKGROUND: European Member States, the Europe...   2020-10-01   \n",
       "4  OBJECTIVE: A key challenge towards a successfu...   2021-06-15   \n",
       "\n",
       "                                             authors            journal  \\\n",
       "0  Zhang, Shiqi; Wang, Xiaobo; Zhang, Hong; Xu, A...     Reprod Toxicol   \n",
       "1                                       Thum, Thomas        Eur Heart J   \n",
       "2  Zhang, Yong; Mao, Dailing; Roswit, William T; ...        Nat Immunol   \n",
       "3  Belfroid, Evelien; RoÎ²kamp, Dorothee; Fraser, ...  BMC Public Health   \n",
       "4  Sethi, Sonika; Kumar, Aditi; Mandal, Anandadee...           BMJ Open   \n",
       "\n",
       "   mag_id  who_covidence_id  arxiv_id  \\\n",
       "0     NaN               NaN       NaN   \n",
       "1     NaN               NaN       NaN   \n",
       "2     NaN               NaN       NaN   \n",
       "3     NaN               NaN       NaN   \n",
       "4     NaN               NaN       NaN   \n",
       "\n",
       "                                      pdf_json_files  \\\n",
       "0  document_parses/pdf_json/04d0cf9898016ae1295e1...   \n",
       "1  document_parses/pdf_json/bc35bf4d3debb1af7ff8d...   \n",
       "2  document_parses/pdf_json/8ff5c024beea4259b7b1b...   \n",
       "3  document_parses/pdf_json/8a61e2af9bdd4ba0b83be...   \n",
       "4  document_parses/pdf_json/7448479341487a112b8e4...   \n",
       "\n",
       "                                 pmc_json_files  \\\n",
       "0  document_parses/pmc_json/PMC7286226.xml.json   \n",
       "1  document_parses/pmc_json/PMC7239191.xml.json   \n",
       "2  document_parses/pmc_json/PMC4653074.xml.json   \n",
       "3  document_parses/pmc_json/PMC7527265.xml.json   \n",
       "4  document_parses/pmc_json/PMC8210694.xml.json   \n",
       "\n",
       "                                                 url        s2_id  \n",
       "0  https://api.elsevier.com/content/article/pii/S...  219560235.0  \n",
       "1  https://www.ncbi.nlm.nih.gov/pubmed/32383758/;...  218557404.0  \n",
       "2  https://www.ncbi.nlm.nih.gov/pubmed/26479788/;...   19031101.0  \n",
       "3  https://doi.org/10.1186/s12889-020-09307-y; ht...  222071121.0  \n",
       "4  https://doi.org/10.1136/bmjopen-2021-048856; h...  235450913.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_path = f'{data_path}/metadata.csv'\n",
    "\n",
    "meta_df = pd.read_csv(metadata_path, dtype={\n",
    "    'pubmed_id': str,\n",
    "    'Microsoft Academic Paper ID': str,\n",
    "    'doi': str\n",
    "})\n",
    "\n",
    "meta_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Currently, the novel coronavirus disease (COVID-19) has outbroken in China and exponentially spread along the world since the first case was diagnosed on December 2019 in Wuhan City of China [1]. The pathogenic novel coronavirus (SARS-CoV-2), isolated from the nasal and pharyngeal secretion, was highly homologous with the coronavirus caused Severe Acute Respiratory Syndrome (SARS) [2]. Owing to the lack of immunity this global pandemic increases sharply within a few months, which not only burns a large amount of health expenditure but seriously threatens lives.<br>The amputating of COVID-19 transmission route is one of the effective measures to prevent the continuing spread. This disease is believed to transmit by inhalation or contact with infected droplets, which is confirmed by the detection of coronavirus in nasopharyngeal swabs [3] and saliva [4] of COVID-19 patients. Apart from the respiratory tract, the SARS-CoV-2 could be also detected in urine and gastrointestinal tract (positive percentage: urine: 0.076%; stool: 0.30%; rectal swab: 0.30%) [5]. Yet, to the best of our knowledge the detection of SARS-CoV-2 in the prostate or expressed prostatic secretion (EPS) has not been reported up till now.<br>Angiotensin converting enzyme 2 (ACE2) was firstly found as a modulator involved in the regulation of renin-angiotensin aldosterone system (RAAS) [6]. It distributes specifically in alveolar epithelial cells, enterocytes of the small intestine and testis in SARS pathogenesis [7]. ACE2 bearing cells are found to be more vulnerable to the attack of coronavirus since the coronavirus spike protein mediates the viral entry to its target cells [8,9]. Indeed, ACE2 has been proved to be a putative receptor for COVID-19 recently [8]. This probably explains the pathogenesis of lung and intestine during SARS and COVID-19. The prostate is also one of the organs express abundant ACE2 [10]. Therefore, it is reasonable to hypothesize that the prostate may be affected by SARS-CoV-2. In this case male fertility could be influenced, and sexual transmission also needs to be taken into account. Yet, recent studies have found the absence of SARS-CoV-2 in the semen and testis [11,12] although the testis is an ACE rich organ. Hence, we aimed to delineate in this study whether SARS-CoV-2 could be detected in EPS and if so, whether COVID-19 was transmitted via sex behavior. Our study was also designed to further prove the initial finding concerning the absence of the virus in semen as the EPS is an essential part of it. The clinical features of these patients were also analyzed, with the focus on the inflammatory indicators and serum SARS-CoV-2 immunoglobins.<br>A total of 10 male inpatients with confirmed diagnosis of NCP were recruited from quarantine wards in the Cancer Center, Union Hospital, Tongji Medical College, Huazhong University of Science and Technology, Wuhan, 430022, China.<br>The diagnosis of COVID-19 was confirmed by either 1) the real-time reverse-transcriptase polymerase-chain-reaction (RT-PCR) or sequencing for nasopharyngeal swabbing SARS-CoV-2 RNA [13], or 2) the positive serum immunoglobin M (IgM) or IgG antibodies of coronavirus. Patients with NCP was the patients confirmed with both COVID-19 and pneumonia diagnosed by CT scan.<br>Patients over 80 years or concomitant with prostatic cancer were excluded.<br>The maximum temperature was recorded according to the medical history given by each patient, but the pulse, respiration, blood pressure and other routine inspections were recorded at admission. The serological parameters were basically collected the second day after admission.<br>All subjects gave their oral and written informed consent before recruitment. This project is approved by the local ethical committee (No. Quick-PJ2020-03-35) in terms of urgent need.<br>Although the Meares-Stamey 4-glass test is the standard method to obtain EPS, there was a study showed that the pre- and post-massage test has strong concordance with the 4-glass test [14]. Considering the time and the cost, the EPS in our study was obtained pre- and post-palpation of the prostate by digital rectal examination, which is also called âtwo-glass testâ. Abstinence was required for at least 5 days before examination. The two operating doctors had been well trained by the urologists beforehand. The detailed operation included two steps. The initial step was the wearing of protective clothing, including the goggle and protective facial screen, one KN95 facial mask plus one surgical facial mask, three surgical gloves, three shoe covers, one protective suit and one isolation gown. All procedures were supervised by the trained staffs from the Centers for Disease Control (CDC). The second step for the two-glass test was the digital rectal examination. The patients were asked to take a knee chest position after the disinfection of external urethral orifice. The urine was collected hereafter as âthe first glass of urineâ. Afterwards, the doctors inserted the index that lubricated with paraffin oil to the recta and palpated the prostate gently. The EPS was collected as âthe second glass of EPSâ.<br>The RT-PCR kit (DAAN Gene Co, Ltd of Sun Yat-sen University) was used for SARS-CoV-2 RNA detection in EPS. To ensure the consistent positive nasopharyngeal SARS-CoV-2, nasopharyngeal swabbing was retested within 3 days of the EPS taken day. Serum IgM and IgG were also tested.<br>A questionnaire on sociodemographic and sexual behaviors would be conducted if the EPS was positive.<br>Quantitative data were depicted as median and interquartile range (Percentile 25, Percentile 75). Statistical analyses were performed with SPSS26.0.<br>The clinical and demographic data of all patients were depicted as the median and interquartile range in brackets in Table 1\n",
      "and the serological data in the same format were shown in Table 2\n",
      ".<br>The median age of infected patients was 57.50 years old, ranging from 29 to 76 years old as shown in Table 1. All patients had a fever before hospitalization (maximum temperature: 39.00 (38.60, 39.23) â), while patients had slightly increased pulse frequency (96.00 (84.25, 115.50) beats per minute), normal respiratory frequency (20.00 (18.75, 22.00) times per minute) and normal blood oxygen saturation (96.50 (93.50, 99.00) %) when admitted. The median interval time between the emerge of the respiratory symptoms / fever and the date of confirmed diagnosis of NCP was 11.00 (8.25, 17.00) days. None of the patients had medical history of tumor, heart failure, cerebrovascular disease, kidney disease, or nursing house stay. Five out of ten patients had diabetes, two had hypertension and two had Type B hepatitis. As displayed, their blood pressure (systolic blood pressure: 127.50 (115.75, 139.25) mmHg; diastolic blood pressure: 90.50 (69.25, 96.00) mmHg) and liver function (Total bilirubin: 12.75 (8.23, 20.73) Î¼mol/L; Alanine aminotransferase: 36.50 (26.25, 62.50) u/L) were well-controlled. However, the blood glucose level was elevated (HbA1c in diabetic patients: 8.80 (6.70, 9.35) %). In addition, d-dimer and fibrinogen were increased, suggesting hypercoagulation and hyperfibrinolysis in these patients. Neither acute renal dysfunction nor myocardial injury was found in these ten patients (serum creatinine: 88.00 (69.75, 92.75) Î¼mol/L; troponin: 2.25 (1.55, 5.63) ng/L).<br>As to the inflammatory indicators, white blood cell (4.94 (4.06, 5.47) Ã109/L) and lymph cell counts (1.43 (0.57, 1.79) Ã109/L) were both within normal range although the lymph cell counts were relatively lower. C-reaction protein (CRP) (3.28 (1.14, 33.33) mg/L) and erythrocyte sedimentation rate (ESR, 22.50 (8.00, 78.50) mm/h) increased as compared to the normal reference whereas procalcitonin (0.075 (0.065, 0.105) ng/ml) was normal. Likewise, interleukin-6 (IL-6) (6.49 (4.96, 21.09) pg/ml) increased while other humoral immune factors, including IL-2, IL-4, IL-10, interferon-Î³, and tumor necrosis factor-Î± were within normal range. The average ratio of CD4+ cells to CD8+ cells (2.04 (1.58, 2.67)) was slightly raised but still within the range.<br>All patients revealed either unilateral or bilateral ground-glass opacity in lung CT scan. Only one of the patients who had chronic obstructive pulmonary disease (COPD) presented suspicious unilateral pleural effusion.<br>The patients were stratified into two groups subsequently. The first group was the patients who had positive nasopharyngeal swabbing SARS-CoV-2 within 3 days of the EPS taken day (PNS group, n = 3) and the other group was the patients who had positive nasopharyngeal swabbing SARS-CoV-2 before admission but turned negative few days before the EPS taken day (PNNS group, n = 7).<br>Although nasopharyngeal swabbing SARS-CoV-2 RNA had been reported positive as a medical history for all patients before admitted, it turned negative within seven patients in the nasopharyngeal swabbing re-tests before the EPS was taken. Only three out of ten patients had positive swabbing nasopharyngeal SARS-CoV-2 RNA within 3 days. Therefore, the interval between the latest negative SARS-CoV-2 RNA test and the day that EPS was taken was counted for the previous seven patients. The longest interval was 29 days while the latest was 3 days before the EPS test as shown in Table 3\n",
      ".<br>Serum SARS-CoV-2 IgM and IgG were also tested qualitatively during hospitalization (Table 3). Unfortunately, two patients had unspecific tests for SARS-CoV-2 immunoglobin because of the shortage for specific testing kit (shown in Table 3). There were two patients with positive nasopharyngeal SARS-CoV-2 RNA manifested positive serum IgM, while all patients had positive serum IgG (Table 3).<br>Unexpectedly but fortunately, none of the patients had positive COVID-19 RNA in EPS regardless of nasopharyngeal SARS-CoV-2 RNA.<br>All patients in our study had a fever, maximum at 39.6â, before hospitalization which supported the importance of temperature test in public area. Our data also presented increased inflammatory indicators, i.e. CRP, ESR, and IL-6 (Table 1), which has also been previously proved by other studies [15,16]. In our study, serum IgM against SARS-CoV-2 was only positive for patients in the PNS group indicating a recent infection. At present, there is no evidence for how long IgM will sustain. However, our data to some extent signified that serum IgM could be a potential and safer surrogate for nasopharyngeal swabbing SARS-CoV-2 RNA as a recent infectious indicator.<br>It has been raised concerns previously with respect to the male reproductive system during COVID-19 infection since the virus was verified to be related to testis damage [17]. Sexual transmission of COVID-19 has also been concerned since another pandemic pathogen, Ebola virus, was detected in semen and contagious in-between sexual partners [18]. The semen sample was still positive for Ebola virus at 565 days after discharge [18].<br>The EPS is secreted by prostate and is an essential component of semen, accounting for one tenth to one third volume of the ejaculation. Prostate fluid protects and nourishes the sperm cells. It would be emphasized that the EPS was negative for the SARS-CoV-2 RNA test according to our data, notwithstanding it was supposed that the prostate was one of the most susceptible organs. The SARS-CoV-2 RNA in EPS was not only negative in the PNNS group, but also negative in the PNS group. This strongly suggested the absence of SARS-CoV-2 in prostate, which possibly implied no prostate injury and the safety of the sexual behavior during COVID-19. In keeping with our data, several studies published very recently have demonstrated the negativity of SARS-CoV-2 in semen sample [11,12,19] although in vitro study had proposed the susceptibility of testis [17]. On the contrary, another recent publication reported that six COVID-19 patients had positive SARS-CoV-2 in semen samples [20]. However, the method description on the obtaining of semen samples in this publication were not fully described. Even so, this result makes the sexual transmission of SARS-CoV-2 more ambiguous. Given the fact that EPS is an important part of the semen, our results utterly supports these previous negative findings. However, it is worthy to mention that the EPS in our study independently verified the negativity of SARS-CoV-2 in the prostate which is different from semen sample in the previous findings.<br>Nonetheless, it would be prudent to recognize the false negative results in SARS-CoV-2 RNA test although the false negative results were more common in the nasopharyngeal swabbing RNA test in terms of non-standard swabbing operation. The potentially high false negative in the nasopharyngeal swabbing test has been raised concerns of the clinicians lately [21] since there were patients had a âturn positiveâ of COVID-CoV-2 but actually had a false negative test previously [22]. Consequently, RNA tests in repetition are needed both for the nasopharyngeal swabbing and for the EPS test in the future to confirm the true negative. In addition, the relatively small sample size is a major limitation of our study in the light of the strict inclusion criteria. A larger cohort is warranted to confirm the absence of sexual transmission during COVID-19. The copy number of viruses also needs to be assessed in parallel, while Meares-Stamey 4-glass Test might also need to be operated instead of two glass tests in the future study. Another limitation of the study is the no sampling of the seminal fluid. A study with both EPS and seminal fluid is therefore called for confirming the absence of SARS-CoV-2.<br>Considering what we have studied is only a tip of the iceberg, more collative actions and scientific evidences are needed to turn this pandemic.<br>The authors declare that there are no conflicts of interest.<br>This research was funded by the grant from the prevention and control research emergency project of COVID-19, Medical Humanities Research Center, Anhui Medical University (âEmergency management and strategies during COVID-19â YJSK202012 to H.Z.) and the project of critical techniques on emergency rescue site during public emergencies, Anhui Medical University (201302003 to G.F. and A.X.).\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "\n",
    "\n",
    "class Article:\n",
    "    def __init__(self, pmcid):\n",
    "\n",
    "        self.paper_id = ''\n",
    "        self.abstract = []\n",
    "        self.body_text = []\n",
    "\n",
    "\n",
    "        if not isinstance(pmcid, str) and math.isnan(pmcid):\n",
    "            return\n",
    "\n",
    "        with open(f\"{all_data_path}/document_parses/pmc_json/{pmcid}.xml.json\") as file:\n",
    "            content = json.load(file)\n",
    "            content_metadata = meta_df.loc[meta_df['pmcid'] == pmcid]\n",
    "\n",
    "            self.paper_id = content['paper_id']\n",
    "            self.abstract = []\n",
    "            self.body_text = []\n",
    "            self.metadata = {}\n",
    "\n",
    "            if not content_metadata is None:\n",
    "                self.metadata = content_metadata\n",
    "\n",
    "            if 'abstract' in content_metadata:\n",
    "                # Abstract\n",
    "                # self.abstract.append(content_metadata['abstract'][0])\n",
    "                for entry in content_metadata['abstract']:\n",
    "                    self.abstract.append(str(entry))\n",
    "                # print(self.abstract)\n",
    "            # Body text\n",
    "            if 'body_text' in content:\n",
    "                for entry in content['body_text']:\n",
    "                    self.body_text.append(entry['text'])\n",
    "\n",
    "            self.abstract = '<br>'.join(self.abstract)\n",
    "            self.body_text = '<br>'.join(self.body_text)\n",
    "\n",
    "    def __repr__(self):\n",
    "        return f'{self.paper_id}: {self.abstract[:200]}... {self.body_text[:200]}...'\n",
    "\n",
    "\n",
    "first_row = Article(meta_df['pmcid'][0])\n",
    "print(first_row.body_text)\n",
    "# meta_df.iloc[0]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dask import dataframe as dd\n",
    "\n",
    "global partial_df\n",
    "global dict_\n",
    "\n",
    "dict_ = {'paper_id': [], 'abstract': [], 'body_text': []}\n",
    "partial_df = pd.DataFrame(dict_, columns=[\n",
    "    'paper_id', 'abstract', 'body_text'])\n",
    "partial_df = dd.from_pandas(partial_df, npartitions=10)\n",
    "\n",
    "partial_df.compute().to_csv(f'{data_path}/df_covid.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing index: 0 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 1000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 2000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 3000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 4000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 5000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 6000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 7000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 8000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 9000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 10000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 11000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "Processing index: 12000 of 12500\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n",
      "saving current progress\n",
      "reseting partial df\n",
      "reseting partial dict\n"
     ]
    }
   ],
   "source": [
    "\n",
    "def populateDict(content):\n",
    "    # no metadata, skip this paper\n",
    "    if len(content.metadata) == 0:\n",
    "        return\n",
    "\n",
    "    # print(meta_data)\n",
    "\n",
    "    dict_['paper_id'].append(content.paper_id)\n",
    "    dict_['abstract'].append(content.abstract.replace('\\n', '<br>'))\n",
    "    dict_['body_text'].append(content.body_text.replace('\\n', '<br>'))\n",
    "\n",
    "\n",
    "def saveProgress():\n",
    "    global partial_df\n",
    "    global dict_\n",
    "\n",
    "    partial_df = pd.DataFrame(dict_, columns=[\n",
    "        'paper_id', 'abstract', 'body_text'])\n",
    "    partial_df = dd.from_pandas(partial_df, npartitions=1)\n",
    "\n",
    "    print('saving current progress')\n",
    "    partial_df.compute().to_csv(\n",
    "        f'{data_path}/df_covid.csv', mode='a', header=False, index=False)\n",
    "\n",
    "    print('reseting partial df')\n",
    "    del partial_df\n",
    "\n",
    "    print('reseting partial dict')\n",
    "    del dict_\n",
    "    dict_ = {'paper_id': [], 'abstract': [], 'body_text': []}\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(meta_df['pmcid']):\n",
    "    if idx % 1000 == 0:\n",
    "        print(f'Processing index: {idx} of {len(meta_df)}')\n",
    "        saveProgress()\n",
    "\n",
    "    populateDict(Article(entry))\n",
    "\n",
    "saveProgress()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>abstract</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC7286226</td>\n",
       "      <td>due to the cellular entry of the novel coronav...</td>\n",
       "      <td>currently the novel coronavirus disease covid1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC7239191</td>\n",
       "      <td>nan</td>\n",
       "      <td>brthis editorial refers to cell typespecific e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC4653074</td>\n",
       "      <td>enhancing the response to interferon could off...</td>\n",
       "      <td>the interferon signaling pathway is considered...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC7527265</td>\n",
       "      <td>background european member states the european...</td>\n",
       "      <td>crossborder outbreaks demonstrate that in an i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC8210694</td>\n",
       "      <td>objective a key challenge towards a successful...</td>\n",
       "      <td>covid19 is an infectious disease that is cause...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     paper_id                                           abstract  \\\n",
       "0  PMC7286226  due to the cellular entry of the novel coronav...   \n",
       "1  PMC7239191                                                nan   \n",
       "2  PMC4653074  enhancing the response to interferon could off...   \n",
       "3  PMC7527265  background european member states the european...   \n",
       "4  PMC8210694  objective a key challenge towards a successful...   \n",
       "\n",
       "                                           body_text  \n",
       "0  currently the novel coronavirus disease covid1...  \n",
       "1  brthis editorial refers to cell typespecific e...  \n",
       "2  the interferon signaling pathway is considered...  \n",
       "3  crossborder outbreaks demonstrate that in an i...  \n",
       "4  covid19 is an infectious disease that is cause...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dask import dataframe as dd\n",
    "import re\n",
    "\n",
    "\n",
    "def lower_case(input_str):\n",
    "    input_str = input_str.lower()\n",
    "    return input_str\n",
    "\n",
    "\n",
    "df_covid = dd.read_csv(f'{data_path}/df_covid.csv')\n",
    "\n",
    "df_covid['body_text'] = df_covid['body_text'].astype(str)\n",
    "df_covid['abstract'] = df_covid['abstract'].astype(str)\n",
    "\n",
    "df_covid['abstract'] = df_covid['abstract'].apply(\n",
    "    lambda x: re.sub('[^a-zA-z0-9\\s]', '', x), meta=('abstract', 'str'))\n",
    "df_covid['body_text'] = df_covid['body_text'].apply(\n",
    "    lambda x: re.sub('[^a-zA-z0-9\\s]', '', x), meta=('body_text', 'str'))\n",
    "\n",
    "df_covid['abstract'] = df_covid['abstract'].apply(\n",
    "    lambda x: lower_case(x), meta=('abstract', 'str'))\n",
    "df_covid['body_text'] = df_covid['body_text'].apply(\n",
    "    lambda x: lower_case(x), meta=('body_text', 'str'))\n",
    "\n",
    "df_covid.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>abstract</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PMC7286226</td>\n",
       "      <td>due to the cellular entry of the novel coronav...</td>\n",
       "      <td>currently the novel coronavirus disease covid1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PMC7239191</td>\n",
       "      <td>nan</td>\n",
       "      <td>brthis editorial refers to cell typespecific e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PMC4653074</td>\n",
       "      <td>enhancing the response to interferon could off...</td>\n",
       "      <td>the interferon signaling pathway is considered...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PMC7527265</td>\n",
       "      <td>background european member states the european...</td>\n",
       "      <td>crossborder outbreaks demonstrate that in an i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PMC8210694</td>\n",
       "      <td>objective a key challenge towards a successful...</td>\n",
       "      <td>covid19 is an infectious disease that is cause...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     paper_id                                           abstract  \\\n",
       "0  PMC7286226  due to the cellular entry of the novel coronav...   \n",
       "1  PMC7239191                                                nan   \n",
       "2  PMC4653074  enhancing the response to interferon could off...   \n",
       "3  PMC7527265  background european member states the european...   \n",
       "4  PMC8210694  objective a key challenge towards a successful...   \n",
       "\n",
       "                                           body_text  \n",
       "0  currently the novel coronavirus disease covid1...  \n",
       "1  brthis editorial refers to cell typespecific e...  \n",
       "2  the interferon signaling pathway is considered...  \n",
       "3  crossborder outbreaks demonstrate that in an i...  \n",
       "4  covid19 is an infectious disease that is cause...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_covid.to_csv(f'{data_path}/df_covid_preprocessed.csv', single_file=True, compute=True, index=False)\n",
    "\n",
    "df_covid.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>abstract</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paper_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC1054884</th>\n",
       "      <td>viral recombination can dramatically impact ev...</td>\n",
       "      <td>as increasing numbers of fulllength viral sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1110908</th>\n",
       "      <td>a wide range of rna viruses use programmed 1 r...</td>\n",
       "      <td>severe acute respiratory syndrome sars first a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1149493</th>\n",
       "      <td>just as proteins form distinct structural moti...</td>\n",
       "      <td>hepatitis delta virus hdv is a satellite virus...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1160249</th>\n",
       "      <td>genstyle  is a workspace designed for the char...</td>\n",
       "      <td>a great number of dna sequences are now availa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1324949</th>\n",
       "      <td>background tuberculosis which is caused by myc...</td>\n",
       "      <td>onethird of the worlds population is estimated...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     abstract  \\\n",
       "paper_id                                                        \n",
       "PMC1054884  viral recombination can dramatically impact ev...   \n",
       "PMC1110908  a wide range of rna viruses use programmed 1 r...   \n",
       "PMC1149493  just as proteins form distinct structural moti...   \n",
       "PMC1160249  genstyle  is a workspace designed for the char...   \n",
       "PMC1324949  background tuberculosis which is caused by myc...   \n",
       "\n",
       "                                                    body_text  \n",
       "paper_id                                                       \n",
       "PMC1054884  as increasing numbers of fulllength viral sequ...  \n",
       "PMC1110908  severe acute respiratory syndrome sars first a...  \n",
       "PMC1149493  hepatitis delta virus hdv is a satellite virus...  \n",
       "PMC1160249  a great number of dna sequences are now availa...  \n",
       "PMC1324949  onethird of the worlds population is estimated...  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import dask.dataframe as dd\n",
    "\n",
    "df_covid = dd.read_csv(f'{data_path}/df_covid_preprocessed.csv').set_index('paper_id')\n",
    "\n",
    "# df_covid = df_covid.drop(\n",
    "#     [\"Unnamed: 0\", \"authors\", \"journal\"], axis=1)\n",
    "\n",
    "\n",
    "df_covid['body_text'] = df_covid['body_text'].astype(str)\n",
    "df_covid['abstract'] = df_covid['abstract'].astype(str)\n",
    "\n",
    "df_covid.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paragraph</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paper_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC1054884</th>\n",
       "      <td>viral recombination can dramatically impact ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1110908</th>\n",
       "      <td>a wide range of rna viruses use programmed 1 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1149493</th>\n",
       "      <td>just as proteins form distinct structural moti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1160249</th>\n",
       "      <td>genstyle  is a workspace designed for the char...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1324949</th>\n",
       "      <td>background tuberculosis which is caused by myc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    paragraph\n",
       "paper_id                                                     \n",
       "PMC1054884  viral recombination can dramatically impact ev...\n",
       "PMC1110908  a wide range of rna viruses use programmed 1 r...\n",
       "PMC1149493  just as proteins form distinct structural moti...\n",
       "PMC1160249  genstyle  is a workspace designed for the char...\n",
       "PMC1324949  background tuberculosis which is caused by myc..."
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "df_covid['body_text'] = df_covid['body_text'].apply(\n",
    "    lambda x: x.split('<br>'), meta=('abstract', 'str'))\n",
    "\n",
    "abstract_df = df_covid.drop(['body_text'], axis=1).replace('nan', np.nan).dropna(subset=['abstract'])\n",
    "df_sentences = abstract_df.rename(columns={'abstract': 'paragraph'})\n",
    "\n",
    "body_text_df = df_covid.drop(['abstract'], axis=1).rename(columns={'body_text': 'paragraph'})\n",
    "df_sentences = df_sentences.append(body_text_df.explode('paragraph'))\n",
    "\n",
    "# df_sentences = df_sentences.replace('NaN', np.nan).dropna(subset=['paragraph'])\n",
    "\n",
    "df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paragraph</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paper_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC1054884</th>\n",
       "      <td>viral recombination can dramatically impact ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1110908</th>\n",
       "      <td>a wide range of rna viruses use programmed 1 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1149493</th>\n",
       "      <td>just as proteins form distinct structural moti...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1160249</th>\n",
       "      <td>genstyle  is a workspace designed for the char...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1324949</th>\n",
       "      <td>background tuberculosis which is caused by myc...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    paragraph\n",
       "paper_id                                                     \n",
       "PMC1054884  viral recombination can dramatically impact ev...\n",
       "PMC1110908  a wide range of rna viruses use programmed 1 r...\n",
       "PMC1149493  just as proteins form distinct structural moti...\n",
       "PMC1160249  genstyle  is a workspace designed for the char...\n",
       "PMC1324949  background tuberculosis which is caused by myc..."
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sentences.to_csv(f'{data_path}/covid_sentences.csv', single_file=True, compute=True)\n",
    "\n",
    "df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# text_dict = text.to_dict()\n",
    "# len_text = len(text_dict[\"paper_id\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paper_id_list = []\n",
    "# body_text_list = []\n",
    "\n",
    "# title_list = []\n",
    "# abstract_list = []\n",
    "# abstract_summary_list = []\n",
    "# for i in tqdm(range(0, len_text)):\n",
    "#     paper_id = text_dict[\"paper_id\"][i]\n",
    "#     body_text = text_dict[\"body_text\"][i].split(\"<br>\")\n",
    "#     title = text_dict[\"title\"][i]\n",
    "#     abstract = text_dict[\"abstract\"][i]\n",
    "#     abstract_summary = text_dict[\"abstract_summary\"][i]\n",
    "#     for b in body_text:\n",
    "#         paper_id_list.append(paper_id)\n",
    "#         body_text_list.append(b)\n",
    "#         title_list.append(title)\n",
    "#         abstract_list.append(abstract)\n",
    "#         abstract_summary_list.append(abstract_summary)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_sentences = pd.DataFrame({\"paper_id\": paper_id_list}, index=body_text_list)\n",
    "# df_sentences.to_csv(f'{root_path}/covid_sentences_body.csv')\n",
    "# df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from dask import dataframe as dd\n",
    "\n",
    "# df_sentences = pd.DataFrame({\"paper_id\": paper_id_list, \"title\": title_list,\n",
    "#                             \"abstract\": abstract_list, \"abstract_summary\": abstract_summary_list}, index=body_text_list)\n",
    "# df_sentences = dd.from_pandas(df_sentences)\n",
    "# df_sentences.to_csv(f'{root_path}/covid_sentences.csv')\n",
    "# df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paragraph</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>paper_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PMC1054884</th>\n",
       "      <td>viral recombination can dramatically impact ev...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1054884</th>\n",
       "      <td>as increasing numbers of fulllength viral sequ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1110908</th>\n",
       "      <td>a wide range of rna viruses use programmed 1 r...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1110908</th>\n",
       "      <td>severe acute respiratory syndrome sars first a...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMC1149493</th>\n",
       "      <td>just as proteins form distinct structural moti...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    paragraph\n",
       "paper_id                                                     \n",
       "PMC1054884  viral recombination can dramatically impact ev...\n",
       "PMC1054884  as increasing numbers of fulllength viral sequ...\n",
       "PMC1110908  a wide range of rna viruses use programmed 1 r...\n",
       "PMC1110908  severe acute respiratory syndrome sars first a...\n",
       "PMC1149493  just as proteins form distinct structural moti..."
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from dask import dataframe as dd\n",
    "\n",
    "df_sentences = dd.read_csv(f'{data_path}/covid_sentences.csv', blocksize=32e6).set_index('paper_id')#.rename(columns={'Unnamed: 0': 'index'})\n",
    "\n",
    "df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_sentences = df_sentences.set_index(\"Unnamed: 0\")\n",
    "\n",
    "# df_sentences.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_covid.to_csv(f'{root_path}/covid_sentences.csv', single_file=True, compute=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_sentences = df_sentences[\"paper_id\"].to_dict()\n",
    "# df_sentences_list = list(df_sentences.keys())\n",
    "# len(df_sentences_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# list(df_sentences.keys())[:5]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_sentences_list = [str(d) for d in tqdm(df_sentences_list)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/UKPLab/sentence-transformers/blob/master/examples/application_semantic_search.py\n",
    "\"\"\"\n",
    "This is a simple application for sentence embeddings: semantic search\n",
    "We have a corpus with various sentences. Then, for a given query sentence,\n",
    "we want to find the most similar sentence in this corpus.\n",
    "This script outputs for various queries the top 5 most similar sentences in the corpus.\n",
    "\"\"\"\n",
    "\n",
    "import torch\n",
    "from sentence_transformers import SentenceTransformer\n",
    "\n",
    "# embedder = SentenceTransformer('multi-qa-MiniLM-L6-cos-v1')\n",
    "embedder = SentenceTransformer(f'{all_data_path}/models/pretrained/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# embedder.save(f'{root_path}/models/pretrained/', 'multi-qa-MiniLM-L6-cos-v1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Batches: 100%|ââââââââââ| 720/720 [03:28<00:00,  3.46it/s]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Corpus with example sentences\n",
    "corpus = df_sentences['paragraph']\n",
    "\n",
    "# corpus.compute()[1]\n",
    "\n",
    "if not os.path.exists(f'{data_path}/corpus_embeddings.npy'):\n",
    "    corpus_embeddings = embedder.encode(\n",
    "        corpus.compute(), device='cuda', show_progress_bar=True)\n",
    "\n",
    "    torch.save(corpus_embeddings, f'{data_path}/corpus_embeddings.npy')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pandas as pd\n",
    "\n",
    "# df = pd.read_csv(f'{root_path}/covid_sentences.csv', index_col=0)\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "paragraph    hepatitis delta virus hdv is a satellite virus...\n",
       "Name: PMC1149493, dtype: object"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NOT SCALABLE\n",
    "df_sentences = df_sentences.compute()\n",
    "\n",
    "# df_sentences.iloc[5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = corpus.loc[corpus['ind'] == 0].compute()\n",
    "\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Batches: 100%|ââââââââââ| 1/1 [00:00<00:00, 46.42it/s]\n"
     ]
    }
   ],
   "source": [
    "import scipy.spatial\n",
    "import torch\n",
    "\n",
    "corpus_embeddings = torch.load(f'{data_path}/corpus_embeddings.npy')\n",
    "\n",
    "# Query sentences:\n",
    "queries = ['What has been published about medical care?',\n",
    "           'Knowledge of the frequency, manifestations, and course of extrapulmonary manifestations of COVID-19, including, but not limited to, possible cardiomyopathy and cardiac arrest',\n",
    "           'Use of AI in real-time health care delivery to evaluate interventions, risk factors, and outcomes in a way that could not be done manually',\n",
    "           'Resources to support skilled nursing facilities and long term care facilities.',\n",
    "           'Mobilization of surge medical staff to address shortages in overwhelmed communities .',\n",
    "           'Age-adjusted mortality data for Acute Respiratory Distress Syndrome (ARDS) with/without other organ failure â particularly for viral etiologies .']\n",
    "query_embeddings = embedder.encode(queries, device='cuda', show_progress_bar=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Top 5 most similar sentences in corpus:\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== What has been published about medical care? =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.6148) \n",
      "\n",
      "Article Index:    9078 \n",
      "\n",
      "Paragraph:    the mass availability and use of mobile health mhealth technology provide a significant potential for such technologies to be integrated into clinical \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5991) \n",
      "\n",
      "Article Index:    15993 \n",
      "\n",
      "Paragraph:    in india assaults on medical workforce such as doctors have been happening for a long time the covid19 epidemic brought the violence against doctors o \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5728) \n",
      "\n",
      "Article Index:    16396 \n",
      "\n",
      "Paragraph:    patient experience of nursing care is correlated with safety clinical effectiveness care quality treatment outcomes including mortality and overall se \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5612) \n",
      "\n",
      "Article Index:    7913 \n",
      "\n",
      "Paragraph:    dear editorbrfirstly we would like to thank hsieh et al [1]brfor their insightful and necessary exploration into the application of appreciative inqui \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5516) \n",
      "\n",
      "Article Index:    17088 \n",
      "\n",
      "Paragraph:    professional medical associations have been challenged to reexamine their purpose and portfolio of value because of significant and accelerating chang \n",
      "\n",
      "-------------------------------------------\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== Knowledge of the frequency, manifestations, and course of extrapulmonary manifestations of COVID-19, including, but not limited to, possible cardiomyopathy and cardiac arrest =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.6830) \n",
      "\n",
      "Article Index:    10983 \n",
      "\n",
      "Paragraph:    background of all patients infected with covid19 95 have mild symptoms but 5 may experience severe illness there are reports of myocardial injury asso \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6614) \n",
      "\n",
      "Article Index:    5896 \n",
      "\n",
      "Paragraph:    dietcontrolled hyperlipidemiabrnonebracute coronary syndrome covid19 viral myocarditis pulmonary embolismbrinitial electrocardiogram ekg showed stemi  \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6602) \n",
      "\n",
      "Article Index:    17009 \n",
      "\n",
      "Paragraph:    1 background patients suffering from the novel coronavirus 2019 covid19 disease could experience several extrapulmonary involvements including cardiov \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6545) \n",
      "\n",
      "Article Index:    10061 \n",
      "\n",
      "Paragraph:    clinical manifestations of covid19 are varied in the general population this study aimed to systematize the literature regarding clinical manifestatio \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6516) \n",
      "\n",
      "Article Index:    17681 \n",
      "\n",
      "Paragraph:    viral infections such as coronavirus disease 2019 covid19 can trigger respiratory infections that may lead to considerable lung damage and even death  \n",
      "\n",
      "-------------------------------------------\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== Use of AI in real-time health care delivery to evaluate interventions, risk factors, and outcomes in a way that could not be done manually =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.6896) \n",
      "\n",
      "Article Index:    22033 \n",
      "\n",
      "Paragraph:    artificial intelligence ai has seen dramatic growth over the past decade evolving from a niche super specialty computer application into a powerful to \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6197) \n",
      "\n",
      "Article Index:    13128 \n",
      "\n",
      "Paragraph:    leveraging artificial intelligence ai approaches in animal health ah makes it possible to address highly complex issues such as those encountered in q \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6184) \n",
      "\n",
      "Article Index:    13129 \n",
      "\n",
      "Paragraph:    artificial intelligence ai encompasses a large range of theories and technologies used to solve problems of high logical or algorithmic complexity it  \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.6140) \n",
      "\n",
      "Article Index:    9128 \n",
      "\n",
      "Paragraph:    this paper focuses on the use of black box ai in medicine and asks whether the physician needs to disclose to patients that even the best ai comes wit \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5922) \n",
      "\n",
      "Article Index:    15172 \n",
      "\n",
      "Paragraph:    the first few months of 2020 have profoundly changed the way we live our lives and carry out our daily activities although the widespread use of futur \n",
      "\n",
      "-------------------------------------------\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== Resources to support skilled nursing facilities and long term care facilities. =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.4871) \n",
      "\n",
      "Article Index:    4859 \n",
      "\n",
      "Paragraph:    at the onset of a disaster or when many unknowns exist it is not clear how many healthcare staff will be needed healthcare administrators and emergenc \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.4702) \n",
      "\n",
      "Article Index:    6039 \n",
      "\n",
      "Paragraph:    the purpose of this article is to present a guide for determining whether a facilitybrhas adequate and appropriate nurse staffing for background we re \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.4683) \n",
      "\n",
      "Article Index:    6040 \n",
      "\n",
      "Paragraph:    us nursing homes are required to have sufficient nursing staff with the appropriate competencies to assure resident safety and attain or maintain the  \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.4634) \n",
      "\n",
      "Article Index:    16426 \n",
      "\n",
      "Paragraph:    individuals transition into longterm care ltc homes closer to the end of their life often with multiple comorbid conditions higher levels of frailty a \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.4566) \n",
      "\n",
      "Article Index:    19048 \n",
      "\n",
      "Paragraph:    the novel coronavirus has shaken the walls of our health care infrastructure in the united states as unprecedented economic social and political facto \n",
      "\n",
      "-------------------------------------------\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== Mobilization of surge medical staff to address shortages in overwhelmed communities . =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.6016) \n",
      "\n",
      "Article Index:    2865 \n",
      "\n",
      "Paragraph:    the tragedy of september 11 anthrax attacks and severe acute respiratory syndrome sars and other recent infectious disease outbreaks have heightened o \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5918) \n",
      "\n",
      "Article Index:    5846 \n",
      "\n",
      "Paragraph:    key stakeholders were identified and a core steering team was constructed with the following structure physician representation from critical care med \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5583) \n",
      "\n",
      "Article Index:    4859 \n",
      "\n",
      "Paragraph:    at the onset of a disaster or when many unknowns exist it is not clear how many healthcare staff will be needed healthcare administrators and emergenc \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5576) \n",
      "\n",
      "Article Index:    4858 \n",
      "\n",
      "Paragraph:    in disasters such as the covid19 pandemic we need to use all available resources to bolster our healthcare workforce many factors go into this process \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.5559) \n",
      "\n",
      "Article Index:    5848 \n",
      "\n",
      "Paragraph:    patients included 541 critically ill and mechanically ventilated patients over a 2year period 20142015 280 and 261 respectively age ranged from 19 to  \n",
      "\n",
      "-------------------------------------------\n",
      "\n",
      "\n",
      "=========================================================\n",
      "==========================Query==============================\n",
      "=== Age-adjusted mortality data for Acute Respiratory Distress Syndrome (ARDS) with/without other organ failure â particularly for viral etiologies . =====\n",
      "=========================================================\n",
      "Score:    (Score: 0.7580) \n",
      "\n",
      "Article Index:    16517 \n",
      "\n",
      "Paragraph:    objective this study aimed to identify the predictors and threshold of failure in neonatal acute respiratory distress syndrome methods newborns with s \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.7270) \n",
      "\n",
      "Article Index:    2162 \n",
      "\n",
      "Paragraph:    there is a striking difference in the reported occurrence of acute respiratory distress syndrome ards between critically ill children and adults while \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.7182) \n",
      "\n",
      "Article Index:    4226 \n",
      "\n",
      "Paragraph:    since the initial description of acute respiratory distress syndrome ards in 1967 [1] its definition has been clarified by expert consensus [2 3] and  \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.7011) \n",
      "\n",
      "Article Index:    19510 \n",
      "\n",
      "Paragraph:    acute respiratory distress syndrome ards is caused by an acute inflammatory injury to the lung associated with increased pulmonary vascular permeabili \n",
      "\n",
      "-------------------------------------------\n",
      "Score:    (Score: 0.7000) \n",
      "\n",
      "Article Index:    18221 \n",
      "\n",
      "Paragraph:    as of 22 june 2021 there have been 3876081 losses of human life reported worldwide due to the covid19 pandemic [1] covid19 pneumonia resulting in crit \n",
      "\n",
      "-------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "def get_corpus_row(idx):\n",
    "    row = df_sentences.iloc[idx]\n",
    "\n",
    "    return row\n",
    "\n",
    "# Find the closest 5 sentences of the corpus for each query sentence based on cosine similarity\n",
    "closest_n = 5\n",
    "print(\"\\nTop 5 most similar sentences in corpus:\")\n",
    "for query, query_embedding in zip(queries, query_embeddings):\n",
    "    distances = scipy.spatial.distance.cdist(\n",
    "        [query_embedding], corpus_embeddings, \"cosine\")[0]\n",
    "\n",
    "    results = zip(range(len(distances)), distances)\n",
    "    results = sorted(results, key=lambda x: x[1])\n",
    "\n",
    "    print(\"\\n\\n=========================================================\")\n",
    "    print(\"==========================Query==============================\")\n",
    "    print(\"===\", query, \"=====\")\n",
    "    print(\"=========================================================\")\n",
    "\n",
    "    for idx, distance in results[0:closest_n]:\n",
    "        print(\"Score:   \", \"(Score: %.4f)\" % (1-distance), \"\\n\")\n",
    "        print(\"Article Index:   \", idx, \"\\n\")\n",
    "        print(\"Paragraph:   \", get_corpus_row(idx)['paragraph'][:150], \"\\n\")\n",
    "        # row_dict = df_sentences.loc[df_sentences.index ==\n",
    "        #                             corpus.loc[idx]].to_dict()\n",
    "        # print(\"paper_id:  \", row_dict[\"paper_id\"][corpus[idx]], \"\\n\")\n",
    "        # print(\"Title:  \", row_dict[\"title\"][corpus[idx]], \"\\n\")\n",
    "        # print(\"Abstract:  \", row_dict[\"abstract\"][corpus[idx]], \"\\n\")\n",
    "        # print(\"Abstract_Summary:  \",\n",
    "        #       row_dict[\"abstract_summary\"][corpus[idx]], \"\\n\")\n",
    "        print(\"-------------------------------------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "\n",
    "torch.cuda.is_available()\n",
    "\n",
    "torch.cuda.current_device()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "TODO: make a custom file (seekable) containing a list of embedings and it's corresponding paper id\n",
    "if possible: make it a data structure where finding the knn is < n^2, possibilities are quad-trees \n",
    "if possible: make it cuda optimized\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "a2ea2f29b5b448882058d44cbab5bc411432e8b9448117a1eba11f607af628be"
  },
  "kernelspec": {
   "display_name": "Python 3.9.7 64-bit ('39venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
